File: data_structures/linked_list/SinglyLinkedList.java
package data_structures.linked_list;

public class SinglyLinkedList<E> {
    private Node<E> head = null;
    private Node<E> tail = null;
    private int size = 0;

    public static class Node<E> {
        public E element;
        public Node<E> next;

        Node(E value) {
            this(value, null);
        }

        Node(E value, Node<E> next) {
            this.element = value;
            this.next = next;
        }
    }

    public SinglyLinkedList(E value) {
        head = new Node<E>(value, null);
        tail = head;
        size++;
    }

    public SinglyLinkedList() {
        head = null;
        tail = null;
    }

    public E getFirst() {
        return head != null ? head.element : null;
    }

    public E getLast() {
        return tail != null ? tail.element : null;
    }

    public int getSize() {
        return size;
    }

    public E addFirst(E element) {
        head = new Node<E>(element, head);

        if(tail == null)
            tail = head;
        size++;

        return head.element;
    }

    public E addLast(E element) {
        E added = null;

        if(tail != null) {
            tail.next = new Node<E>(element);
            tail = tail.next;
            added = tail.element;
            size++;
        }

        return added == null ? addFirst(element) : added;
    }

    public E removeFirst() {
        Node<E> tmp = head;

        if(isEmpty())
            return null;

        if(head == tail)
            tail = null;

        E removed = head.element;
        head = head.next;
        tmp.next = null;
        size--;

        return removed;
    }

    public E removeLast() {
        Node<E> tailPrev;
        Node<E> tmp = head;

        if(isEmpty())
            return null;

        while(tmp.next != tail && tmp.next != null) {
            tmp = tmp.next;
        }

        tailPrev = tmp;
        tmp = tail;
        E removed = tail.element;
        if(tailPrev == tmp) {
            tmp.next = null;
            tail = null;
            head = null;
        } else {
            tail = tailPrev;
            tail.next = null;
        }

        size--;

        return removed;
    }

    public E insertAt(E element, int position) {
        Node<E> tmp = head;

        if(position < 1 || head == null)
            return null;

        for(int i = 1; i < position && i <= size; i++) {
            if(tmp.next != null)
                tmp = tmp.next;
        }

        Node<E> next = position == 1 ? tmp : tmp.next;
        Node<E> nodeToInsert = new Node<>(element, next);
        if(position == 1) {
            nodeToInsert.next = head;
            head = nodeToInsert;
        } else if(position >= getSize()) {
            tail = nodeToInsert;
            tmp.next = nodeToInsert;
        } else {
            tmp.next = nodeToInsert;
        }
        size++;

        return tmp.next.element;
    }

    public E removeAt(int position) {
        int count = 1;
        Node<E> tmp = head;

        if(position < 1)
            return null;

        for(int i = 1; i < position && i <= size; i++) {
            if(tmp.next != null)
                tmp = tmp.next;
        }

        Node<E> removed = tmp.next;
        tmp.next = removed != null ? removed.next : null;
        size--;

        return removed != null ? removed.element : null;
    }

    public void clear() {
        Node<E> tmp = head;

        while(tmp != null) {
            tmp = tmp.next;
            head.next = null;
            head = tmp;
        }

        tail = null;
        size = 0;
    }

    public E find(E elementToFind) {
        Node<E> tmp = head;

        while(tmp != null) {
            if(tmp.element.equals(elementToFind))
                return tmp.element;
            tmp = tmp.next;
        }

        return null;
    }

    public boolean contains(E elementToCheck) {
        return find(elementToCheck) != null;
    }

    public boolean isEmpty() {
        return size == 0;
    }
}

File: data_structures/linked_list/singly_linked_list.cpp
class SinglyLinkedList {
    struct Node {
        int value;
        Node *next;
        Node(int val) : value(val), next(0) {}
    };
    Node *head;
    Node *tail;
    int size;
public:
    SinglyLinkedList()
        : head(0), tail(0), size(0) {}
    SinglyLinkedList(int value) : size(1)
        { head = new Node(value); tail = head; }
    SinglyLinkedList(const SinglyLinkedList&) = delete;


    ~SinglyLinkedList();

    SinglyLinkedList& operator=(const SinglyLinkedList&) = delete;
    int Search(int value) const;
    void PushFront(int value);
    void PushBack(int value);
    int PopFront();
    int PopBack();
    bool InsertAt(int value, int pos);
    bool RemoveAt(int position);
    int Size() const;
};

SinglyLinkedList::~SinglyLinkedList()
{
    Node *tmp = head;
    while(tmp) {
        head = head->next;
        delete tmp;
        tmp = head;
    }
}

int SinglyLinkedList::Search(int value) const
{
    const Node *tmp;
    int position = 0;

    for(tmp = head; tmp; tmp = tmp->next) {
        if(tmp->value == value)
            return position;
        position++;
    }

    return -1;
}

void SinglyLinkedList::PushFront(int value)
{
    Node *tmp = new Node(value);
    tmp->next = head;
    head = tmp;
    size++;
}

void SinglyLinkedList::PushBack(int value)
{
    Node *tmp = new Node(value);
    if(!head) {
        head = tail = tmp;
    } else {
        tail->next = tmp;
        tail = tmp;
    }
    size++;
}

int SinglyLinkedList::PopFront()
{
    if(!head)
        return 0;

    Node *tmp = head;
    int val = tmp->value;

    head = head->next;
    if(!head)
        tail = 0;

    delete tmp;
    size--;

    return val;
}

int SinglyLinkedList::PopBack()
{
    if(!tail)
        return 0;

    int val = tail->value;

    if(head == tail) {
        delete head;
        head = tail = 0;
    } else {
        Node *tmp = head;
        while(tmp->next != tail)
            tmp = tmp->next;

        delete tail;
        tail = tmp;
        tail->next = 0;
    }

    size--;

    return val;
}

bool SinglyLinkedList::InsertAt(int value, int pos)
{
    if(pos > size || pos < 0)
        return false;

    Node *new_node = new Node(value);
    Node *tmp = head;
    Node *tmp_prev = 0;

    for(int i = 0; i < pos; i++) {
        tmp_prev = tmp;
        tmp = tmp->next;
    }

    new_node->next = tmp;
    if(tmp_prev)
        tmp_prev->next = new_node;
    else
        head = new_node;

    if(new_node->next == 0)
        tail = new_node;

    size++;

    return true;
}

bool SinglyLinkedList::RemoveAt(int position)
{
    if(position >= size || position < 0)
        return false;

    Node *tmp = head;
    Node *tmp_prev = 0;

    for(int i = 0; i < position; i++) {
        tmp_prev = tmp;
        tmp = tmp->next;
    }

    if(tmp_prev) {
        tmp_prev->next = tmp->next;
        if(!tmp->next)
            tail = tmp_prev;
    } else {
        head = tmp->next;
        if(!head)
            tail = 0;
    }

    delete tmp;
    size--;

    return true;
}

int SinglyLinkedList::Size() const
{
    return size;
}
